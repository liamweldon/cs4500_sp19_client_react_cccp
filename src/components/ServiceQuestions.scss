.questions {
  button {
    background: none;
    border: none;
  }
  i {
    font-size: 2em;

    &.fa-trash-alt {
      color: red;
    }
    &.fa-pen-square,
    &.fa-plus-square {
      color: blue;
    }
    &.fa-check-square {
      color: green;
    }
  }
}
