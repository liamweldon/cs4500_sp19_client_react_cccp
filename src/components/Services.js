import React from "react";
import "./table.scss";

const Services = ({
  services,
  service,
  findAllServices,
  createService,
  deleteService,
  selectService,
  updateService,
  updateNameInForm,
  updateDescriptionInForm
}) => (
  <div>
    <h3>Services</h3>
    <table className="table">
      <tbody>
        <tr key={"header"}>
          <td>
            <div className="row">
              <div className="col-5">
                <h5>Service Name</h5>
              </div>
              <div className="col-5">
                <h5>Description</h5>
              </div>
            </div>
          </td>
        </tr>
        <tr key={"edit"}>
          <td>
            <div className="row">
              <div className="col-5">
                <input
                  onChange={e => updateNameInForm(e)}
                  value={service.serviceName}
                />
              </div>
              <div className="col-5">
                <input
                  onChange={e => updateDescriptionInForm(e)}
                  value={service.serviceDescription}
                />
              </div>
              <div className="col-1">
                <button
                  onClick={createService}
                  style={{ backgroundColor: "blue", borderColor: "blue" }}
                >
                  <i
                    className="fa fa-plus"
                    style={{ color: "white", fontSize: "50px" }}
                  />
                </button>
              </div>
              <div className="col-1">
                <button
                  onClick={updateService}
                  style={{ backgroundColor: "green", borderColor: "green" }}
                >
                  <i
                    className="fa fa-check"
                    style={{ color: "white", fontSize: "50px" }}
                  />
                </button>
              </div>
            </div>
          </td>
        </tr>
        {services.map(serv => (
          <tr className="serviceRow" key={serv.id}>
            <td>
              <div className="row">
                <div className="col-5">
                  <a href={"/admin/services/" + serv.id}>{serv.serviceName}</a>
                </div>
                <div className="col-5">
                  <a href={"/admin/services/" + serv.id}>
                    {serv.serviceDescription}
                  </a>
                </div>
                <div className="col-1">
                  <button
                    className={"delete-button" + serv.id}
                    onClick={() => deleteService(serv.id)}
                    style={{ backgroundColor: "red", borderColor: "red" }}
                  >
                    <i
                      className="fa fa-times"
                      style={{ color: "white", fontSize: "50px" }}
                    />
                  </button>
                </div>
                <div className="col-1">
                  <button
                    onClick={() => selectService(serv.id)}
                    style={{ backgroundColor: "blue", borderColor: "blue" }}
                  >
                    <i
                      className="fa fa-edit"
                      style={{ color: "white", fontSize: "50px" }}
                    />
                  </button>
                </div>
              </div>
            </td>
          </tr>
        ))}
      </tbody>
    </table>
  </div>
);

export default Services;
