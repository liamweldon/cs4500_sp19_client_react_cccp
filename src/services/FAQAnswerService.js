import {API_ROOT} from '../api-config';

export default class FAQAnswerService {
    static instance = null;
    static getInstance() {
        if (FAQAnswerService.instance === null) {
            FAQAnswerService.instance = new FAQAnswerService();
        }
        return this.instance;
    }
    findFAAById = (id) => fetch(`${API_ROOT}/api/faas/${id}`).then((response) => response.json());
    findAllFAAs = () => fetch(`${API_ROOT}/api/faas`).then((response) => response.json());
    addFAA = (newAnswer) =>
        fetch(`${API_ROOT}/api/faas/`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify(newAnswer)
        }).then((response) => response.json());

    editFAA = (id, newAnswer) =>
        fetch(`${API_ROOT}/api/faas/${id}`, {
            method: 'PUT',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify(newAnswer)
        }).then((response) => response.json());
}