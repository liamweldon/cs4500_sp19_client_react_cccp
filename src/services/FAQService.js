import {API_ROOT} from '../api-config';

export default class FAQService {
  static instance = null;
  static getInstance() {
    if (FAQService.instance === null) {
      FAQService.instance = new FAQService();
    }
    return this.instance;
  }

  addFAQ = (newQuestion) =>
    fetch(`${API_ROOT}/api/faqs/`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(newQuestion)
    }).then((response) => response.json());

  editFAQ = (id, newQuestion) =>
    fetch(`${API_ROOT}/api/faqs/${id}`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(newQuestion)
    }).then((response) => response.json());
  findFAQById = (id) => fetch(`${API_ROOT}/api/faqs/${id}`).then((response) => response.json());
  deleteFAQ = (id) => fetch(`${API_ROOT}/api/faqs/${id}`, {method: 'DELETE'}).then((response) => response.json());
  findAllFAQs = () => fetch(`${API_ROOT}/api/faqs`).then((response) => response.json());
  linkFAAtoFAQ = (faaID, faqID) => fetch(`${API_ROOT}/api/faqs/${faqID}/faas/${faaID}`).then(response => response.json());
}
