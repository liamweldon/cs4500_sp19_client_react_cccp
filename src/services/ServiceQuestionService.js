import {API_ROOT} from '../api-config';

export default class ServiceQuestionService {
    static instance = null;
    static getInstance() {
        if (ServiceQuestionService.instance === null) {
            ServiceQuestionService.instance = new ServiceQuestionService();
        }
        return this.instance;
    }
    findServiceQuestionById = (id) =>
        fetch(`${API_ROOT}/api/service-questions/${id}`).then((response) => response.json());

    findAllServiceQuestions = () => fetch(`${API_ROOT}/api/service-questions`).then((response) => response.json());

    addServiceQuestion = (newServiceQuestion) =>
    fetch(`${API_ROOT}/api/service-questions/`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(newServiceQuestion)
    }).then((response) => response.json());
}
