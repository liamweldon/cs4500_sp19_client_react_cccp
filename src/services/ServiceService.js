import { API_ROOT } from "../api-config";

export default class ServiceService {
  static instance = null;

  static getInstance() {
    if (ServiceService.instance === null) {
      ServiceService.instance = new ServiceService();
    }
    return this.instance;
  }

  findServiceById = serviceId =>
    fetch(`${API_ROOT}/api/services/${serviceId}/`).then(response =>
      response.json());

  findAllServices = () =>
    fetch(`${API_ROOT}/api/services`).then(response => response.json());

  createService = service => {
    delete service.id;
    return fetch(`${API_ROOT}/api/services`,
    {
      method: 'post',
      body: JSON.stringify(service),
      headers: {'content-type': 'application/json'}
    })
    .then(response => response.json())
  };

  updateService = service => {
    return fetch(`${API_ROOT}/api/services/${service.id}`,
    {
      method: 'put',
      body: JSON.stringify(service),
      headers: {'content-type': 'application/json'}
    })
    .then(response => response.json())
  };

  deleteService = id => {
    return fetch(`${API_ROOT}/api/services/${id}`,
    {
      method: 'delete'
    })
  };

}
