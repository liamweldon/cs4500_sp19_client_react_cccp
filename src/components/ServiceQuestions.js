import React from 'react'
import ServiceQuestionService from '../services/ServiceQuestionService'
import './table.scss';
import {Link} from 'react-router-dom';
class ServiceQuestions extends React.Component {
    constructor(props) {
        super(props)
        this.serviceQuestionService = ServiceQuestionService.getInstance()
        this.state = {
            serviceQuestions: [],
            newServiceQuestion: {
              serviceType: '',
              type: '',
              question: ''
            }
        }
    }

    getServiceQuestions = () => {
      this.serviceQuestionService.findAllServiceQuestions().then((questions) =>
        this.setState({
          serviceQuestions: questions
        })
      );
    };

    componentDidMount() {
        this.serviceQuestionService
            .findAllServiceQuestions()
            .then(serviceQuestions =>
                this.setState({
                    serviceQuestions: serviceQuestions
                })
            )
    }

    handleNewServiceQuestionInputChange = (e) => {
      this.setState({newServiceQuestion: {...this.state.newServiceQuestion, [e.target.name]: e.target.value}});
    };

    addServiceQuestion = () => {
      const newServiceQuestion = this.state.newServiceQuestion;
      this.ServiceQuestionService.addServiceQuestion(newServiceQuestion).then((res) => {
        this.getServiceQuestions();
        this.setState({
          newServiceQuestion: {
          serviceType: '',
          type: '',
          question: ''
        }
      });
    });
  };

    render() {
        return(
            <div>
                <h3>Service Questions</h3>
                <table className="table">
                    <tbody>
                    <tr className="header-row">
                            <td>Type</td>
                            <td>Question</td>
                    </tr>
                    <tr key={-1}>
                      <td>
                        <input value={this.state.newServiceQuestion.service} name="service" onChange={this.handleNewServiceQuestionInputChange} />
                      </td>
                      <td>
                        <select
                          value={this.state.newServiceQuestion.type}
                          name="type"
                          onChange={this.handleNewServiceQuestionInputChange}
                        />
                      </td>
                      <td>
                        <input
                          value={this.state.newServiceQuestion.question}
                          name="question"
                          onChange={this.handleNewServiceQuestionInputChange}
                        />
                      </td>


                    </tr>
                    {this.state.serviceQuestions.map((serviceQuestion) => (
                            <tr key={serviceQuestion.id}>
                                <td>
                                    {serviceQuestion.type}
                                </td>
                                <td><Link to={`/admin/service-questions/${serviceQuestion.id}`}>{serviceQuestion.question}</Link></td>
                            </tr>
                        ))}
                    </tbody>
                </table>
            </div>
        )
    }
}

export default ServiceQuestions
