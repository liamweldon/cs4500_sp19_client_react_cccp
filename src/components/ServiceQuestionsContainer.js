import React from 'react'
import ServiceQuestionService from '../services/ServiceQuestionService';
import ServiceService from "../services/ServiceService";
import ServiceQuestions from './ServiceQuestions.js';
import './table.scss';
import {Link} from 'react-router-dom';
import {omit} from 'lodash';

class ServiceQuestionsContainer extends React.Component {
    constructor(props) {
        super(props)
        this.serviceQuestionService = ServiceQuestionService.getInstance()
        this.serviceService = ServiceService.getInstance();
        this.state = {
            serviceQuestions: [],
            services: [],
            editing: {},
            newServiceQuestion: {
              serviceName: '',
              type: '',
              question: ''
            }
        }
    }

    getServiceQuestions = () => {
      this.serviceQuestionService.findAllServiceQuestions().then((questions) =>
        this.setState({
          serviceQuestions: questions
        })
      );
    };

    componentDidMount() {
        this.serviceQuestionService
            .findAllServiceQuestions()
            .then(serviceQuestions =>
                this.setState({
                    serviceQuestions: serviceQuestions
                })
            );
        this.serviceService
            .findAllServices()
            .then(services =>
                this.setState({
                    services: services
                })
            );
    }

      deleteQuestion = (id) => {
        // once the question is deleted, reload the questions so the user doesnt see the deleted one
        this.serviceQuestionService.deleteServiceQuestion(id).then((res) => this.getServiceQuestions());
      };

    saveServiceQuestion = (id) => {
      const newServiceQuestion = this.state.editing[id];
      this.serviceQuestionService.editServiceQuestion(id, newServiceQuestion).then((res) => {
        this.getServiceQuestions();
        this.setState({editing: omit(this.state.editing, id)});
      });
    };

    handleNewServiceQuestionInputChange = (e) => {
      this.setState({newServiceQuestion: {...this.state.newServiceQuestion, [e.target.name]: e.target.value}});
    };


    addServiceQuestion = () => {
      const newServiceQuestion = this.state.newServiceQuestion;
      this.serviceQuestionService.addServiceQuestion(newServiceQuestion).then((res) => {
        this.getServiceQuestions();
        this.setState({
          newServiceQuestion: {
            serviceName: '',
            type: '',
            question: ''
        }
      });
    });
  };

  handleInputChange = (e, id) => {
    let editingCopy = {...this.state.editing};
    editingCopy[id] = {...editingCopy[id], [e.target.name]: e.target.value};
    this.setState({editing: editingCopy});
  };

  toggleEditMode = (serviceQuestion) => {
    if (serviceQuestion.id in this.state.editing) {
      this.setState({editing: omit(this.state.editing, serviceQuestion.id)});
    } else {
      this.setState({
        editing: {...this.state.editing, [serviceQuestion.id]: {
          serviceName: serviceQuestion.serviceName,
          type: serviceQuestion.type,
          question: serviceQuestion.question}
        }
      });
    }
  };

  render() {
    const eventHandlers = {
      handleNewServiceQuestionInputChange: this.handleNewServiceQuestionInputChange,
      handleInputChange: this.handleInputChange,
      handleCreateClick: this.addServiceQuestion,
      handleDeleteClick: this.deleteQuestion,
      handleEditClick: this.toggleEditMode,
      handleSaveClick: this.saveServiceQuestion
    }

    return (
      <ServiceQuestions
        serviceQuestions={this.state.serviceQuestions}
        services={this.state.services}
        editing={this.state.editing}
        newServiceQuestion={this.state.newServiceQuestion}
        eventHandlers={eventHandlers}
      />

    );
  }
}

export default ServiceQuestionsContainer;
