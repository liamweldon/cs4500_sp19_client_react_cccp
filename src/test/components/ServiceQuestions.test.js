import React from 'react'
import renderer from 'react-test-renderer';
import ServiceQuestions from '../../components/ServiceQuestions'
import ServiceQuestionsContainer from '../../components/ServiceQuestionsContainer'
import ServiceQuestionService from '../../services/ServiceQuestionService'
import Services from '../../components/Services'
import ServiceService from '../../services/ServiceService'
import {StaticRouter} from 'react-router';
import Enzyme, {mount} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import serviceQuestions from '../mockData/service-questions.mock.json'
import mockServiceQuestion from '../services/mocks/MockServiceQuestionService'
import services from '../mockData/services.mock.json'
import mockService from '../services/mocks/MockServiceService'

Enzyme.configure({adapter: new Adapter()});

// Baseline initial service question
let mock_initial_service_question = {id: 0, serviceName: "Service Name", type: "Type", question: "Question"}

// Mock both service and service question services
const serviceQuestionService = ServiceQuestionService.getInstance();
const serviceService = ServiceService.getInstance();
beforeEach(() => {
	global.fetch = mockServiceQuestion(serviceQuestions);
})

test('Service Question Summary Snapshot and DOM', () => {
  const component = renderer.create(
    <StaticRouter location="/admin/service-questions" context={{}}>
			<ServiceQuestions serviceQuestions={serviceQuestions} services={services}
				editing={{}} eventHandlers={{}} newServiceQuestion={mock_initial_service_question}/>
    </StaticRouter>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

	let testInstance = component.root;
	let rows = testInstance.findAllByProps({className : "serviceQuestionRow"});
	expect(rows.length).toBe(4);
});
