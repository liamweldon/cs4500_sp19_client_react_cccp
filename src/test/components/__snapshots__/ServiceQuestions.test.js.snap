// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Service Question Summary Snapshot and DOM 1`] = `
<div>
  <h3>
    Service Questions
  </h3>
  <table
    className="table"
  >
    <tbody>
      <tr
        className="header-row"
      >
        <td>
          Service
        </td>
        <td>
          Type
        </td>
        <td>
          Question
        </td>
      </tr>
      <tr>
        <td>
          <select
            name="serviceName"
            value="Service Name"
          >
            <option
              value="House Cleaners"
            >
              House Cleaners
            </option>
            <option
              value="Interior Painters"
            >
              Interior Painters
            </option>
            <option
              value="Math Tutoring"
            >
              Math Tutoring
            </option>
          </select>
        </td>
        <td>
          <select
            name="type"
          >
            <option
              value="Choice"
            >
              Choice
            </option>
            <option
              value="True/False"
            >
              True/False
            </option>
            <option
              value="Range"
            >
              Range
            </option>
          </select>
        </td>
        <td>
          <input
            name="question"
            value="Question"
          />
        </td>
        <td>
          <button>
            Create +
          </button>
        </td>
      </tr>
      <tr
        className="serviceQuestionRow"
      >
        <td>
          Pet Sitting
        </td>
        <td>
          True/False
        </td>
        <td>
          <a
            href="/admin/service-questions/1"
            onClick={[Function]}
          >
            Can you take care of both cats and dogs?
          </a>
        </td>
        <td />
        <td>
          <button
            onClick={[Function]}
          >
            <i
              className="fas fa-trash-alt"
            />
          </button>
        </td>
        <td>
          <i
            className="fas fa-pen-square"
            onClick={[Function]}
          />
        </td>
      </tr>
      <tr
        className="serviceQuestionRow"
      >
        <td>
          Fitness
        </td>
        <td>
          Choice
        </td>
        <td>
          <a
            href="/admin/service-questions/2"
            onClick={[Function]}
          >
            Do you even lift?
          </a>
        </td>
        <td />
        <td>
          <button
            onClick={[Function]}
          >
            <i
              className="fas fa-trash-alt"
            />
          </button>
        </td>
        <td>
          <i
            className="fas fa-pen-square"
            onClick={[Function]}
          />
        </td>
      </tr>
      <tr
        className="serviceQuestionRow"
      >
        <td>
          Housekeeping
        </td>
        <td>
          Choice
        </td>
        <td>
          <a
            href="/admin/service-questions/3"
            onClick={[Function]}
          >
            Do you clean houses or apartments?
          </a>
        </td>
        <td />
        <td>
          <button
            onClick={[Function]}
          >
            <i
              className="fas fa-trash-alt"
            />
          </button>
        </td>
        <td>
          <i
            className="fas fa-pen-square"
            onClick={[Function]}
          />
        </td>
      </tr>
      <tr
        className="serviceQuestionRow"
      >
        <td>
          Handymen
        </td>
        <td>
          True/False
        </td>
        <td>
          <a
            href="/admin/service-questions/4"
            onClick={[Function]}
          >
            Do you own a hammer?
          </a>
        </td>
        <td />
        <td>
          <button
            onClick={[Function]}
          >
            <i
              className="fas fa-trash-alt"
            />
          </button>
        </td>
        <td>
          <i
            className="fas fa-pen-square"
            onClick={[Function]}
          />
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
